ouroboros-network/src/Ouroboros/Network/TxSubmission/Inbound.hs

```diff
@@ -192,16 +226,22 @@ txSubmissionInbound _tracer maxUnacked TxSubmissionMempoolWriter{..} =
                 -> ServerState txid tx
                 -> Collect txid tx
                 -> m (ServerStIdle n txid tx m ())
-    handleReply n st (CollectTxIds reqNo txids) =
+    handleReply n st (CollectTxIds reqNo txids) = do
       -- Upon receiving a batch of new txids we extend our available set,
       -- and extended the unacknowledged sequence.
-      return $ serverIdle n st {
+      --
+      -- We also pre-emptively acknowledge those txids that are already in the
+      -- mempool. This prevents us from requesting their corresponding
+      -- transactions again in the future.
+      let st' = st {
         requestedTxIdsInFlight = requestedTxIdsInFlight st - reqNo,
         unacknowledgedTxIds    = unacknowledgedTxIds st
                               <> Seq.fromList (map fst txids),
         availableTxids         = availableTxids st
                               <> Map.fromList txids
       }
+      mpSnapshot <- atomically mempoolGetSnapshot
+      serverIdle n (acknowledgeTxIdsInMempool st' mpSnapshot)
```

Comment:
I'm sure you and @dcoutts have discussed this already, but just checking again: is this the right and *only* place to look at the Mempool?

Reply:
In my discussions with Duncan, he noted that this seemed like the right place to do it, yeah. I also believe this is where it makes the most sense as well.